source/common/upstream/health_checker_base_impl.cc

```diff
@@ -267,6 +271,13 @@ void HealthCheckerImplBase::ActiveHealthCheckSession::handleSuccess(bool degrade
     // it to healthy. This makes startup faster with a small reduction in overall reliability
     // depending on the HC settings.
     if (first_check_ || ++num_healthy_ == parent_.healthy_threshold_) {
+      // If the host moves to healthy, clear active HC timeout, which may be toggled off and on
+      // while the host is unhealthy.
+      host_->healthFlagClear(Host::HealthFlag::ACTIVE_HC_TIMEOUT);
+      // A host that was told to exclude based on immediate failure, but is now passing, should
+      // no longer be excluded.
+      host_->healthFlagClear(Host::HealthFlag::EXCLUDED_VIA_IMMEDIATE_HC_FAIL);
```

Comment:
I think there is a larger question as to whether the behavior of respecting x-envoy-immediate-health-check-fail should be configurable given the new untrusted upstream model, but IMO that is out of scope for this PR since it wasn't configurable previously for data plane responses. But to answer your specific question, I don't think there would be any substantial difference between a host doing the flopping on the data plane vs. doing the flopping purely with health checks, so if you are OK with it I think we can defer to general hardening in a future PR if we are worried about this path given the new threat model?